pkg/dataobj/decoder.go

```diff
@@ -1,55 +1,102 @@
 package dataobj
 
 import (
+	"bytes"
 	"context"
 	"errors"
 	"fmt"
+	"io"
 
 	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/filemd"
 	"github.com/grafana/loki/v3/pkg/dataobj/internal/util/bufpool"
 )
 
+const optimisticReadBytes = 16 * 1024
+
 type decoder struct {
-	rr rangeReader
+	rr   rangeReader
+	size int64
 }
 
 func (d *decoder) Metadata(ctx context.Context) (*filemd.Metadata, error) {
-	tailer, err := d.tailer(ctx)
+	objectSize, err := d.objectSize(ctx)
 	if err != nil {
-		return nil, fmt.Errorf("reading tailer: %w", err)
+		return nil, fmt.Errorf("reading object size: %w", err)
 	}
 
-	rc, err := d.rr.ReadRange(ctx, int64(tailer.FileSize-tailer.MetadataSize-8), int64(tailer.MetadataSize))
+	readSize := min(objectSize, optimisticReadBytes)
+	buf := bufpool.Get(int(readSize))
+	defer bufpool.Put(buf)
+
+	if err := d.readLastBytes(ctx, readSize, buf); err != nil {
+		return nil, fmt.Errorf("reading last bytes: %w", err)
+	}
+
+	tailer, err := d.tailer(ctx, buf)
 	if err != nil {
-		return nil, fmt.Errorf("getting metadata: %w", err)
+		return nil, fmt.Errorf("reading tailer: %w", err)
+	}
+
+	var rc io.ReadCloser
+	if tailer.MetadataSize+8 > uint64(buf.Len()) {
+		// Our optimistic read was too small, so we need to read the metadata fully
+		rc, err = d.rr.ReadRange(ctx, int64(tailer.FileSize-tailer.MetadataSize-8), int64(tailer.MetadataSize))
+		if err != nil {
+			return nil, fmt.Errorf("getting metadata: %w", err)
+		}
+		defer rc.Close()
+	} else {
+		rc = io.NopCloser(bytes.NewReader(buf.Bytes()[buf.Len()-int(tailer.MetadataSize)-8 : buf.Len()-8]))
 	}
-	defer rc.Close()
 
 	br := bufpool.GetReader(rc)
 	defer bufpool.PutReader(br)
 
 	return decodeFileMetadata(br)
 }
 
+func (d *decoder) readLastBytes(ctx context.Context, readSize int64, buf *bytes.Buffer) error {
+	objectSize, err := d.objectSize(ctx)
+	if err != nil {
+		return fmt.Errorf("reading object size: %w", err)
+	}
+
+	rc, err := d.rr.ReadRange(ctx, objectSize-readSize, readSize)
+	if err != nil {
+		return fmt.Errorf("reading last %d bytes: %w", readSize, err)
+	}
+	defer rc.Close()
+
+	_, err = io.Copy(buf, rc)
```

Comment:
Should we use [io.CopyN](https://pkg.go.dev/io#CopyN) here to make sure we read exactly `readSize`?